architecture/2025-01-serverless-feature-spec.md

```diff
@@ -0,0 +1,279 @@
+
+Users can specify Radius [application](https://docs.radapp.io/reference/resource-schema/core-schema/application-schema/) definitions for applications running on serverless compute platforms. 
+> Must be sure to include serverless platform specific customizations via a `runtimes` property, similar to how Kubernetes patching was implemented for [containers](https://docs.radapp.io/reference/resource-schema/core-schema/container-schema/#runtimes).
+
+Allow for Radius abstraction of a [container](https://docs.radapp.io/reference/resource-schema/core-schema/container-schema/) resource that can be deployed to serverless compute platforms. Must be sure to include serverless platform specific configurations via [connections](https://docs.radapp.io/reference/resource-schema/core-schema/container-schema/#connections), [extensions](https://docs.radapp.io/reference/resource-schema/core-schema/container-schema/#extensions), and routing to other resources via [gateways](https://docs.radapp.io/reference/resource-schema/core-schema/gateway/).
+> One future idea to explore is whether we can we build extensibility via Recipes - i.e. allow Recipes for Containers, which themselves can be serverless containers.
+
+### Scenario 2: "Punch-through" to platform-specific features and incremental adoption of Radius into existing serverless applications
+<!-- One or two sentence summary -->
+Allow for platform-specific features to be used in Radius applications via abstraction "punch-through" mechanisms, similar to how Kubernetes-specific features are supported in Radius via [base YAML or PodSpec patching](https://docs.radapp.io/guides/author-apps/containers/overview/#kubernetes) functionalities that is achieved through the [`runtimes`](https://docs.radapp.io/reference/resource-schema/core-schema/container-schema/#runtimes) property in the container definition. For example, ACI has a confidential containers offering that may not be common across serverless platforms, but Radius should allow for this feature to be used in ACI containers.
+> Stretch goal: Ability to add Radius to existing serverless applications without requiring a full rewrite of the application, similar to how Radius can be added to existing Kubernetes applications via [Kubernetes manifest](https://docs.radapp.io/tutorials/add-radius/) or [Helm chart](https://docs.radapp.io/tutorials/helm/) annotations. Note: The Kubernetes/Helm support works because Kubernetes itself is extensible. Other systems like ACA are not extensible in the same way but we should explore if there are options to make this work.
+
+### Scenario 3: User interfaces for serverless--Radius API, CLI, Dashboard
+<!-- One or two sentence summary -->
+Enable deployment and management of serverless compute resources via the existing Radius API and CLI commands. Serverless resources that are modeled in Radius should be available in the App Graph and Dashboard for visualization and management.
```

Comment:
Are we going to have a scenario where a user can have applications running on Kubernetes clusters and Serverless platforms? I am talking about an app.bicep file with containers with different runtimes.

Reply:
Combination of both worlds.

Reply:
This will not be a functionality for serverless, at least not in the initial implementation. We'll require a single compute platform per environment, which means all containers within the same application must run in the same compute platform for a single deployment. I'll note this in the spec. This has the benefit of simplifying this serverless feature significantly but we also don't expect users to combine compute platforms for a single application as things like cross-platform DNS, ingress, etc. are challenging.

Reply:
I mean I think we don't need to put in extra effort for this because Radius will already be deployed in a Kubernetes cluster and in the app.bicep file we should be able to define resources in both worlds.